### ISSUE TITLE
[NEXT-642] `Link` will not respect `basePath` when used in new `app` directory

### ISSUE THREAD

**Issue by fwcontact** [vercel/next.js#45686]
### Verify canary release

- [X] I verified that the issue exists in the latest Next.js canary release

### Provide environment information

    Operating System:
      Platform: darwin
      Arch: arm64
      Version: Darwin Kernel Version 22.3.0: Thu Jan  5 20:48:54 PST 2023; root:xnu-8792.81.2~2/RELEASE_ARM64_T6000
    Binaries:
      Node: 18.12.1
      npm: 9.4.2
      Yarn: N/A
      pnpm: N/A
    Relevant packages:
      next: 13.1.7-canary.7
      eslint-config-next: 13.1.6
      react: 18.2.0
      react-dom: 18.2.0

### Which area(s) of Next.js are affected? (leave empty if unsure)

App directory (appDir: true), Middleware / Edge (API routes, runtime)

### Link to the code that reproduces this issue

https://github.com/fwcontact/link-bug

### To Reproduce

create pages in `app` dir, then use `Link` to point to other dirs.
the `Link` component can show the url respecting the `basePath`,
but when clicked, the page will not go to the correct url, in effect, missing the `basePath` portion.

### Describe the Bug

create pages in `app` dir, then use `Link` to point to other dirs.
the `Link` component can show the url respecting the `basePath`,
but when clicked, the page will not go to the correct url, in effect, missing the `basePath` portion.

### Expected Behavior

should prefix the url with the `basePath` value

<sub>[NEXT-642](https://linear.app/vercel/issue/NEXT-642/link-will-not-respect-basepath-when-used-in-new-app-directory)</sub>

**Comment by Rishab49** [comment]
can you show your file structure and code

**Comment by fwcontact** [comment]
it’s actually really easy to replicate,
and this is a repo you can use to try out.

https://github.com/fwcontact/link-bug￼

it’s the template project of yours.
the "real deal” is in file `next.config.js`,
where I added:
basePath: '/test',

now when you start the project with `npm run dev`,
you would see 404 page at http://localhost:3000 <http://localhost:3000/>, which is expected.

and then when you goto http://localhost:3000/test, you’ll see the homepage.
at top left, I added 2 links, 
To "/demo1"
To "/demo2"

mouse over them, link url appears to be fine, (in browser status bar)
but when clicking on them, they take you to the wrong place, in effect, “http://localhost:3000/demo1” and “http://localhost:3000/demo2”,
which is of course 404 now. (should have been http://localhost:3000/test/demo1 and http://localhost:3000/test/demo2 respectively)

[link url in status bar, correct]
￼
[clicked on link, navigated to the wrong place]
￼

however, if you manually goto http://localhost:3000/test/demo2
this is a “legacy” page (not using `app` folder)
when you click on the link,
	To "/demo3”
it’s working properly.

**Comment by luanmm** [comment]
I can confirm this behavior. It is happening for me too as I'm migrating a multi-project repo to "app dir".

**Comment by luanmm** [comment]
Seems that the `basePath` is being added only to the `<a>` tag, but not being used to navigate:

1. The URL with `basePath` is being set [here](https://github.com/vercel/next.js/blob/canary/packages/next/src/client/link.tsx#L703) by `addBasePath`
2. The URL used to navigate comes from [this memo](https://github.com/vercel/next.js/blob/canary/packages/next/src/client/link.tsx#L435), where it does not seem to consider the `basePath`

**Comment by mwawrusch** [comment]
Please prioritize this. 

It'd be great if their would be a boolean attribute to indicate whether next/link should honor the base path or not. In most cases it's much safer to just compose the full link ourselves.

**Comment by luanmm** [comment]
Don't know if it is the very same scenario, but for me this was corrected by [this commit](https://github.com/vercel/next.js/commit/1c5983fe95b79cc9ea1e52d19e69adf7827eda5a), related to [this issue](https://github.com/vercel/next.js/issues/41824) (which seems to be a slightly different case, but perhaps for the same reason).

@fwcontact and @mwawrusch, I would test [canary 9](https://github.com/vercel/next.js/releases/tag/v13.1.7-canary.9) and check if it was the same scenario as mine (I think it is). If it works, this issue could be closed in favor of the other one that triggered the (`commit`) action.

**Comment by arshiamalek** [comment]
We are having the same issue and it is related to https://github.com/vercel/next.js/issues/41824. The `Link` component which it shows the correct link when you hover over it but doesn't direct you to the correct URL (It doesn't respect `basePath`). We have tried [v13.1.7-canary.17](https://github.com/vercel/next.js/releases/tag/v13.1.7-canary.17), [v13.1.7-canary.9](https://github.com/vercel/next.js/releases/tag/v13.1.7-canary.9) and [v13.1.7-canary.11](https://github.com/vercel/next.js/releases/tag/v13.1.7-canary.11) and neither of it have fixed our issue.

To reproduce our issue you can follow step in this comment: https://github.com/vercel/next.js/issues/45686#issuecomment-1422259736

**Comment by luanmm** [comment]
@arshiamalek are you using a monorepo or something that could be messing up with the dependencies? The strange thing is that your scenario seems to be exactly what I was having before `v13.1.7-canary.9`.

Just in case, this is the code I'm using in my app `Link` component (for me it is needed because of MUI library):
```tsx
'use client'

import React, { PropsWithRef } from 'react'
import NextLink from 'next/link'
import { Link as MuiLink, type LinkProps } from '@mui/material'

const Link: React.FC<PropsWithRef<LinkProps>> = React.forwardRef(
  (props, ref) => {
    const { href } = props

    return (
      <NextLink href={href as any} legacyBehavior={true} passHref>
        <MuiLink ref={ref} {...props} />
      </NextLink>
    )
  }
)

Link.displayName = 'Link'

export default Link
```

Note that `legacyBehavior` is being set to `true`. I don't know if this make it behave differently as yours, but perhaps it could be the case. I can't test it right now because the project's environment is not set here.

**Comment by teobler** [comment]
Hi @fwcontact I think latest version already fixed your issue, I cannot reproduce it in v13.1.7-canary21.
But here is one more thing we can do to make better development experience:
For now no matter we add basePath or not, console log from cli will always log url as `url: http://localhost:3000`, we can enhance it to something like `url: http://localhost:3000<basePath>`

I can raise a PR for that @balazsorban44

**Comment by teobler** [comment]
Seems `next.config.js` not be imported during `next dev` command, to implement `url: http://localhost:3000<basePath>` we need to import that, not sure if this is a good idea, will pause this idea until maintainer's idea.

**Comment by fwcontact** [comment]
Guys, I just confirmed that the issue has been fixed in the latest version 13.2.1 👍

**Comment by timneutkens** [comment]
Going to close this based on @fwcontact's comment 👍
